### DISCUSSION TITLE
Best Practices for Partitioning GeoParquet?

### DISCUSSION THREAD

**Discussion by cholmes** [opengeospatial/geoparquet#171] [Ecosystem]
I've recently been working with Google's [Open Buildings Dataset](https://sites.research.google/open-buildings/), exploring the creation of a more cloud-native distribution of it than what Google currently offers, likely to put up on [Source Cooperative](https://source.coop/). I've got a PMTiles of the whole dataset, and then I'm also wanting to offer it as GeoParquet. 

My ideal is to let people treat it as a single file, but then have it partitioned so people can also just download / work with say a single country. So was looking for advice / best practices on how to split it up. I think the best example of this thus far is [Microsoft Buildings on Planetary Computer](https://planetarycomputer.microsoft.com/dataset/ms-buildings) So I'm mostly wondering if you could share how that's done @TomAugspurger? How does the delta file work? How did divide up the underlying files? How did you create the delta file? etc.  Also curious about delta vs iceberg vs just using parquet? 

I'm also happy for anyone else who has broken up large datasets to sound in with what you did. Hopefully we can get some best practices and figure out how to make this easy for people. 

The Google Buildings is broken up into S2 chunks, but it does seem like it'd be more useful to have it based on countries. And then maybe break it up further if the amount of data in a country is still quite big? Like I think Nigeria is going to be huge.

**Comment by TomAugspurger** [comment]
> So I'm mostly wondering if you could share how that's done @TomAugspurger?

Fortunately, I didn't have to do that for the version we published through the Planetary Computer. The raw data was in some sort of Spark system the team creating this dataset uses. They exported it to a partitioned parquet dataset. They were able to partition the original dataset by quadkey (something like `df.write.partitionBy("region", "quadkey").parquet("output.parquet")`, which gives the table partitioning described [here](https://spark.apache.org/docs/latest/sql-data-sources-parquet.html#partition-discovery). Notably, their raw data already had the region name and quadkey, which enabled partitioning on those columns. We didn't have to derive it from the polygons.

If I did have to do this myself, I would use Dask. Roughly something like this

```python
df = pd.read_csv("data/*.csv", include_path_column=True) 
# do some regex stuff to extract the S2 chunk ID out of the path
df.to_parquet("data/output.parquet", partition_on="s2_key")
```

but that would need some tweaking to get it to work. Happy to help out with this if needed. 

I think @kylebarron has some code at https://github.com/kylebarron/spatially-partitioned-geoparquet looked into this a bit.

> My ideal is to let people treat it as a single file, but then have it partitioned so people can also just download / work with say a single country.

I agree this is the most convenient way to access the data (though slight nitpick: I'd say "single parquet dataset" rather than "single file", since it'll be partitioned into many files). Systems like Dask, Spark, Synapse, etc. will let you treat the partitioned dataset as a single logical table.

> How does the delta file work? How did divide up the underlying files? How did you create the delta file? etc. 

I don't really understand delta or Synapse, but using it was pretty straightforward:

```python
# raw data from my teammate
df = spark.read.parquet(
    "abfss://footprints@bingmlbuildings.dfs.core.windows.net/global/2023-04-25/ml-buildings.parquet/"
)
# rewritten as delta
df.write.format("delta").partitionBy(["RegionName", "quadkey"]).save(
    "/delta/2023-04-25/ml-buildings.parquet/"
)
```

That spit out the delta format files (which I think is just some kind of JSON metadata file at the root plus the parquet files). 

> How did divide up the underlying files?

We already had good partitioning (at least one file per `(region, quadkey)`; very dense quadkeys were split to keep the files below ~200 MB), so I don't think delta did anything to repartition the files. IIUC, delta was just adding a metadata layer on top.

> Also curious about delta vs iceberg vs just using parquet?

No thoughts on delta vs. iceberg (Synapse has good built-in support for delta, so I used that).

As for delta / iceberg vs. plain parquet, I think that some kind of catalog system can be important. For systems like spark / dask, a `read_parquet("/path/to/dataset.parquet")` can be slow if there are very many files under the root. My teammates had a use-case where that listing was too slow. But with the delta format, they can [query on the partitioned columns](https://planetarycomputer.microsoft.com/dataset/ms-buildings#Example-Notebook) and get super quick access to the matching files. It was pretty impressive just how fast things can be with the right metadata to enable the filtering.

Interestingly, you can *also* (or alternatively) make STAC items for the individual parquet files and get a similar access pattern (see the bottom of https://planetarycomputer.microsoft.com/dataset/ms-buildings#Example-Notebook), but that's another can of worms.

---

Finally, it's worth mentioning that because Synapse / spark / delta don't support writing geoparquet metadata (AFAIK) I have to manually add that. I've been meaning to publish this as a package, but [this gist](https://gist.github.com/TomAugspurger/ae9bc974a5c28e44bc95234c72750c47) has a snippet for adding geoparquet metadata to a parquet file.

**Reply by cholmes** [reply]
Awesome, thanks for all this info - though yes, it seems like the key part that I'm interested in was already mostly done. It'd be awesome if you were able to help me work through this, my first experience with geopandas was working on this. And I've not yet tried out dask (though would be psyched to explore it, and planetary computer more) - I did most all of this on my local machine. I have some questions that are more basic, and hopefully this discussion can be a place to help others who try to do this in the future.

> `df.to_parquet("data/output.parquet", partition_on="s2_key")`

So is this just partitioning 'naively'? Like it just goes through your data and any unique s2_key would get its own file? In the Google Buildings dataset the s2 partitions in the file paths don't seem balanced at all - some are like 2 kb, one is 14 gigabytes. It seems like it might be better to follow the lead of the microsoft team, and try to create columns for region and quadkey. I have the full data in a postgis database now, so should be easy to add country. 

Quadkey I'm less certain on. Just spent a good bit of time digging into the ms buildings dataset, it looks like it's a quadkey at zoom level 9, so that seems fairly straightforward for adding as a column. But I noticed it said 'Each (Region, quadkey) pair will have one or more geoparquet files, depending on the density of the of the buildings in that area.' (and just realized you mentioned this below too) - do you know how this processing was done? Like how was the data split if the quadkey was too dense? And how was a double quadkey of a dense area input to the 'partition_on' call? 

> Systems like Dask, Spark, Synapse, etc. will let you treat the partitioned dataset as a single logical table.

Another naive question - what about GeoPandas? If something does not treat it as a single logical table what happens if you point it at the root file?

> Finally, it's worth mentioning that because Synapse / spark / delta don't support writing geoparquet metadata (AFAIK) I have to manually add that. I've been meaning to publish this as a package, but [this gist](https://gist.github.com/TomAugspurger/ae9bc974a5c28e44bc95234c72750c47) has a snippet for adding geoparquet metadata to a parquet file.

Ah, that's awesome - definitely would be great to have this as a package. Do you know if there's a way to do this 'streaming'? Like add the metadata on the fly on the way out from a cloud bucket? At the geoparquet meetup @jatorre and @alasarr were interested in that kind of functionality.

> `# raw data from my teammate
df = spark.read.parquet(
    "abfss://footprints@bingmlbuildings.dfs.core.windows.net/global/2023-04-25/ml-buildings.parquet/"
)`

Was this just a single big parquet file here?

**Reply by TomAugspurger** [reply]
I started https://github.com/TomAugspurger/google-open-buildings-example. But it won't work for the full thing (unless you have a lot of RAM) because of the issue at https://github.com/TomAugspurger/google-open-buildings-example#issues.  If the `.csv.gz` files were uncompressed you should be able to process the entire thing without running out of memory.

>> df.to_parquet("data/output.parquet", partition_on="s2_key")

> So is this just partitioning 'naively'? Like it just goes through your data and any unique s2_key would get its own file?

Yes, essentially.

> the file paths don't seem balanced at all - some are like 2 kb, one is 14 gigabytes.

This unfortunately also affects the ms-buildings dataset, at least on the small side. I think that's unavoidable for the small files. For larger files, my teammate was able to configure Spark so that the parquet files were a maximum size. So dense regions with many features in a single quadkey would be split up into multiple files.

Roughly, something like:

```
out.parquet/
  s2_cell=093/
      part.0.parquet
      part.1.parquet
```

Sparse regions would have just a single file for that `(region, quadkey)`.

>> Systems like Dask, Spark, Synapse, etc. will let you treat the partitioned dataset as a single logical table.

> Another naive question - what about GeoPandas? If something does not treat it as a single logical table what happens if you point it at the root file?

It looks like geopandas (and pandas) will attempt to read all the files under the root and concatenate them all together into a single (in-memory) DataFrame.

> Do you know if there's a way to do this 'streaming'? Like add the metadata on the fly on the way out from a cloud bucket?

I'm not, unfortunately. Presumably you could write it to a local file system, modify it to add the geoparquet metadata, and then push that to blob storage. I'm not familiar enough with pyspark to pull that off :)

> Was this just a single big parquet file here?

Both the inputs and outputs here were parquet datasets (so the root prefix in blob storage to a whole bunch of parquet files)

<img width="440" alt="image" src="https://github.com/opengeospatial/geoparquet/assets/1312546/250c6521-ee35-4e9a-b4a0-9ca759dfb900">

**Reply by cholmes** [reply]
> I started https://github.com/TomAugspurger/google-open-buildings-example. But it won't work for the full thing (unless you have a lot of RAM) because of the issue at https://github.com/TomAugspurger/google-open-buildings-example#issues.  If the `.csv.gz` files were uncompressed you should be able to process the entire thing without running out of memory.

I spent some time converting the gzipped csv files to filegeodatabase and geoparquet's, seems like using those should help? I also stripped out the lat / long columns, as they seemed redundant. I'm traveling now and they're on a hard disk at home, but I'll modify your script to use them when I get back. I could also upload them to https://source.coop/ - aiming to put the end parquet files there, but could be nice to just have a decently better format that follows the google s2 convention to start.

I managed to get your script working on one building file (all I have local right now), and it helped me figure out how to get my script actually working. I had been on a roll with chatgpt, but the 2021 cut off was an issue here with it not knowing about dask_geopandas. 

Will your script be memory efficient? I probably should just switch to using the cloud, but have been trying to process this stuff locally. It seems like dask is supposed to help with data that doesn't fit in memory (and this is 100gb+ and I've got 32 gigs), but [my script](https://gist.github.com/cholmes/d9ad24efe53fa5e094055ffb3e41979c) right now seems to be memory bound. Next I'm going to try to keep things in dask and only transform to dask_geopandas just before writing out to disk, since that's what you seem to do. 

>> the file paths don't seem balanced at all - some are like 2 kb, one is 14 gigabytes.
> This unfortunately also affects the ms-buildings dataset, at least on the small side. I think that's unavoidable for the small files. 

Gotcha. Ok, I guess that's not the worst thing - especially if you can treat it all as one dataset and this is just details behind it, and we get rid of the 'big' ones.

> For larger files, my teammate was able to configure Spark so that the parquet files were a maximum size. So dense regions with many features in a single quadkey would be split up into multiple files.

Ah, so it's just like some Spark setting? And Spark doesn't yet write geoparquet, right? So to go this route would use Spark to write WKB, then use your script to add the metadata?

>> Another naive question - what about GeoPandas? If something does not treat it as a single logical table what happens if you point it at the root file?

> It looks like geopandas (and pandas) will attempt to read all the files under the root and concatenate them all together into a single (in-memory) DataFrame.

Nice.

>> Do you know if there's a way to do this 'streaming'? Like add the metadata on the fly on the way out from a cloud bucket?

> I'm not, unfortunately. Presumably you could write it to a local file system, modify it to add the geoparquet metadata, and then push that to blob storage. I'm not familiar enough with pyspark to pull that off :)

No worries.

>>  Was this just a single big parquet file here?

> Both the inputs and outputs here were parquet datasets (so the root prefix in blob storage to a whole bunch of parquet files)

<img width="440" alt="image" src="https://github.com/opengeospatial/geoparquet/assets/1312546/250c6521-ee35-4e9a-b4a0-9ca759dfb900">

Got it, thanks!

Thanks for all the help. I'll try to adapt your script to get it working for me.

**Reply by Maxxen** [reply]
Shameless plug, but you could also try DuckDB to manage repartitioning of lots of large parquets files. We support both [partitioned reading with filter-pushdown](https://duckdb.org/docs/data/parquet/overview.html) through [globbing/hive partitioning/union by name](https://duckdb.org/docs/data/parquet/tips.html) and [partitioned writes](https://duckdb.org/docs/data/partitioning/partitioned_writes.html). Although we don't write the geoparquet-specific metadata yet, so you'll have to do that with some sort of post-processing.

**Reply by cholmes** [reply]
Ah, cool - that's a great idea. I've been meaning to try out DuckDB, but haven't had a clear problem for it. I'll probably have a bunch of basic questions about it, but will give it a try and probably start a new discussion. Thanks @maxxen!

**Reply by Maxxen** [reply]
Repartitioning parquet or csv is emerging as a really common use case for DuckDB, feel free to ping me anytime!

**Reply by cholmes** [reply]
Hey @Maxxen - I'm digging in, and first impressions are awesome! Nice interface and really fast. I'm not yet really to the parquet partitioning stuff, mostly am just trying to flex some of the geospatial aspects to get a sense of how I could do things. What's the best forum to ask questions about that? I'm getting some weird performance stuff, like memory going to 60+ gigabytes on two tables that are under 200mb each, using what seems to be a fairly standard spatial join.

**Reply by cholmes** [reply]
I got some first results of partitioning on country and admin-1 with DuckDB, but I'm a bit confused by them.

<img width="1277" alt="Screenshot 2023-06-09 at 3 41 20 PM" src="https://github.com/opengeospatial/geoparquet/assets/407017/bf12bb69-f658-40ba-95b8-8869203e533b">

It did the country and admin level as I expected, but then each admin often has up to 11 files, most of which are 161 bytes - sometimes it has the 161 byte ones, sometimes it just has a handful of the data_4.parquet type files, skipping numbers. If I load a 161 byte one in duck it seems to have the schema but no rows:

```
D select * from 'data_2.parquet';
┌────────────────┬────────────┬────────────────┬─────────────┬─────────┬──────────┐
│ area_in_meters │ confidence │ full_plus_code │ country_iso │ admin_1 │ geometry │
│     double     │   double   │    varchar     │   varchar   │ varchar │   blob   │
├─────────────────────────────────────────────────────────────────────────────────┤
│                                     0 rows                                      │
└─────────────────────────────────────────────────────────────────────────────────┘
```

I'm not sure what it's attempting to divide the data within the file by, and was hoping it'd just be just one file in them (perhaps unless it was a really big file). Any idea @Maxxen? 

My ideal would actually be to have folders for the countries, and then partitions by admin_1, but to have each data file be at least 100mb, and so usually just one file. And even more ideal would be for the name of the file to be the admin_1 value, but I didn't see a way to do that. 

I'm suspecting that my overall goal may not be that compatible with how these partitioning things work. My ideal would be that I could partition into sensible boundaries, like countries and states/provinces. And smart tools could treat the dataset as a single table, but other tools could just navigate to the individual files that users would like. To have a new way of distributing global data, where it'd be partitioned for tools but each piece would be a sensible chunk for people to use. Which is why it would be nice to have the data named something like TUN_Tunis.parquet

**Reply by cholmes** [reply]
Found something as to why it might be writing 11 files, it says 'Currently, one file is written per thread to each directory.' in https://duckdb.org/docs/data/partitioning/partitioned_writes.html

**Reply by Maxxen** [reply]
Hi! I just fixed a memory leak last week which is probably why you're seeing such a high memory usage on joins. We're about to issue a bugfix release today which is going to include a bunch of other fixes and tweaks for the spatial extension as well.

Regarding the partitioning, perhaps I spoke too soon, If you already have the geometries labeled by some partitioning scheme and simply wanted to repartition them by some other column - that should be easy with duckdb regardless of how big the underlying dataset is. But if you want to group by some spatial extent and perform a spatial join, its probably not going to be that fast right now since duckdbs spatial capabilities doesn't use any spatial indexes or geometry-aware joins yet. We basically have to do a n*m loop. Although that might still be acceptable depending on how large each side is, and it should still handle larger-than-memory execution (once the bugfix is released) even if it's less efficient.

I can't really say why you're ending up with the empty data files, but the threads being too aggressive when initialising the partitions are probably a good guess, I've forwarded it as an issue internally and we've been able to reproduce it. We're working on "streaming" partitioning as well so that you could set a limit (e.g. 100mb) before a new datafile is created, but for now you could also set the thread limit with `PRAGMA threads=1` before writing to get one datafile per partition.

**Reply by cholmes** [reply]
> Hi! I just fixed a memory leak last week which is probably why you're seeing such a high memory usage on joins. We're about to issue a bugfix release today which is going to include a bunch of other fixes and tweaks for the spatial extension as well.

Awesome! My favorite type of bug - one that's already been fixed :) Will check it out when it's released.

> Regarding the partitioning, perhaps I spoke too soon, If you already have the geometries labeled by some partitioning scheme and simply wanted to repartition them by some other column - that should be easy with duckdb regardless of how big the underlying dataset is. But if you want to group by some spatial extent and perform a spatial join, its probably not going to be that fast right now since duckdbs spatial capabilities doesn't use any spatial indexes or geometry-aware joins yet.

Ah cool - that was going to be one of my questions, if there was a spatial index yet. Are there plans for it? If you had that plus the ability to write out geoparquet then duckdb will be an incredible tool for the geospatial world, and would likely be my go to tool for the things I'm doing now. I just ended up doing my spatial join in PostGIS and then did the partitioning in DuckDB and it worked great. But I already like working with duckdb much more than postgresql :) It also just blew away using geopandas to read in a big (10 gig) shapefile and write out a flatgeobuf. Did it in a 20th the time and was way easier on my memory.

> I can't really say why you're ending up with the empty data files, but the threads being too aggressive when initialising the partitions are probably a good guess, I've forwarded it as an issue internally and we've been able to reproduce it. 

Great, thanks!

> We're working on "streaming" partitioning as well so that you could set a limit (e.g. 100mb) before a new datafile is created, but for now you could also set the thread limit with PRAGMA threads=1 before writing to get one datafile per partition.

Streaming partitions with a limit sounds great.

**Comment by cholmes** [comment]
Ok, I've finally got something to show on this. Thanks for all the help @TomAugspurger and @Maxxen 
I put my results so far up on source.coop, see https://beta.source.coop/cholmes/google-open-buildings It's a little browser on top of the S3 bucket at s3://us-west-2.opendata.source.coop/google-research-open-buildings/ 
So far there's 2 sets of the data in GeoParquet: 

* The first ([geoparquet/ folder](https://beta.source.coop/cholmes/google-open-buildings/browse/geoparquet) just translates the gzipped csv files into geoparquet (with a bit of clean-up - removing the long/lat columns and splitting multi-polys so it can be all polygon type). This is mostly just to make the original data slightly more accessible, for any experiments in partitioning, etc -  
* The second ([geoparquet-admin1/ folder](https://beta.source.coop/cholmes/google-open-buildings/browse/geoparquet-admin1) I added fields for country (3 character iso field) and admin 1, using the [geoBoundaries](https://www.geoboundaries.org/) dataset. And then I partitioned them using the country field, and named them by their admin_1 field.

The idea of naming them meaningfully appeals to me to make it so the folders could be used in 'download to desktop' type traditional workflows. I also put up a little tutorial on [GeoParquet and DuckDB](https://gist.github.com/cholmes/f78c225f971c6d935e4340ce9895b4b2) as it made it quite easy to do things like just download a whole country of Parquet files and put in a common GIS format. The [httpfs extension](https://duckdb.org/docs/extensions/httpfs.html) made it really easy for someone not all up on S3 to grab the files, and the [spatial extension](https://duckdb.org/docs/extensions/spatial) made it easy to write out to lots of different formats.

GDAL/OGR seemed able to read the custom named files as a partition just fine - I'd be curious if people could test with other tools. I also have no idea if the partitions actually help in the querying all that much.

 If others are interested in experimenting with more partitions it should be a solid dataset to work with, at ~60 gigs. My next steps are to try out using [iceberg](https://iceberg.apache.org/) or [delta](https://docs.databricks.com/delta/index.html) on top of the partitions, and to perhaps try out pure spatial partitioning using [Sedona](https://sedona.apache.org/latest-snapshot/tutorial/sql/#save-geoparquet). Not sure when I'll find the time though, so others are more than welcome to try it out as well. If you get interesting results and want to host them on source.coop let me know and I'm sure we can get it up there.
